lib/linked_list.rb

```diff
@@ -2,44 +2,71 @@
 
 class LinkedList
   attr_reader :head
-
+  
   def initialize
     @head = nil
   end
-
-  # Time complexity - ?
-  # Space complexity - ?
+  
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def add_first(data)
-
+    @head = Node.new(data, head)
   end
-
-  # Time complexity - ?
-  # Space complexity - ?
+  
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def get_first
-
+    return nil if head.nil?
+    return @head.data
   end
-
-  # Time complexity - ?
-  # Space complexity - ?
+  
+  # Time complexity - O(n) where n is the length of the list
+  # Space complexity - O(1)
   def length
-    return 0
+    length_counter = 0
+    current = head
+    until current.nil?
+      current = current.next
+      length_counter += 1
+    end
+    return length_counter
   end
-
-  # Time complexity - ?
-  # Space complexity - ?
+  
+  # Time complexity - O(n) where n is the length of the list
+  # Space complexity - O(1)
   def add_last(data)
-
+    current = @head
+    if current.nil?
+      @head = Node.new(data, nil)
+    else 
+      until current.next.nil?
+        current = current.next
+      end
+      current.next = Node.new(data, nil)
+    end
   end
-
-  # Time complexity - ?
-  # Space complexity - ?
+  
+  # Time complexity - O(n) where n is the length of the list
+  # Space complexity - O(1)
   def get_last
-
+    current = head
```

Comment:
What if `head` is `nil`?

```suggestion
   return nil if head.nil?
    current = head
```